File: projects/order-bc-query/app/src/main/java/com/robinfood/app/usecases/createtransaction/CreateTransactionUseCase.java
package com.robinfood.app.usecases.createtransaction;

import com.robinfood.app.mappers.input.TransactionMappers;
import com.robinfood.core.dtos.request.transaction.RequestTransactionDTO;
import com.robinfood.core.dtos.response.transaction.ResponseTransactionDTO;
import com.robinfood.core.entities.TransactionEntity;
import com.robinfood.core.entities.TransactionFlowEntity;
import com.robinfood.repository.transaction.ITransactionCRUDRepository;
import com.robinfood.repository.transaction.ITransactionRepository;
import com.robinfood.repository.transactionflow.ITransactionFlowRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.concurrent.CompletableFuture;

/**
 * Implementation of ICreateTransactionUseCase
 */
@Component
@Slf4j
public class CreateTransactionUseCase implements ICreateTransactionUseCase {

    private final ITransactionCRUDRepository transactionCRUDRepository;
    private final ITransactionRepository transactionRepository;
    private final ITransactionFlowRepository transactionFlowRepository;

    public CreateTransactionUseCase(
            ITransactionCRUDRepository transactionCRUDRepository,
            ITransactionRepository transactionRepository,
            ITransactionFlowRepository transactionFlowRepository
    ) {
        this.transactionCRUDRepository = transactionCRUDRepository;
        this.transactionRepository = transactionRepository;
        this.transactionFlowRepository = transactionFlowRepository;
    }

    @Override
    public CompletableFuture<ResponseTransactionDTO> invoke(RequestTransactionDTO transactionDTO) {

        log.info("Create transaction has started with request: {}", transactionDTO);

        if (transactionDTO.getId() != null) {
            transactionCRUDRepository.deleteById(transactionDTO.getId());
        }

        final TransactionEntity transactionEntity = TransactionMappers.toTransactionEntity(
                transactionDTO);

        final TransactionEntity savedTransactionEntity = transactionCRUDRepository.save(
                transactionEntity);

        transactionFlowRepository.save(
                new TransactionFlowEntity(
                        savedTransactionEntity.getId(),
                        transactionDTO.getFlowId()));

        if (savedTransactionEntity.getCreatedAt() != null) {
            transactionRepository.setLocalTransaction(savedTransactionEntity);
        }

        savedTransactionEntity.setCreatedAt(transactionRepository.getLocalTransaction().getCreatedAt());

        log.info("Transaction created: {}", savedTransactionEntity);

        return CompletableFuture.completedFuture(
                TransactionMappers.toTransactionDTO(savedTransactionEntity)
        );
    }
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/usecases/createtransaction/CreateTransactionUseCase.kt
package com.robinfood.app.usecases.createtransaction

import com.robinfood.app.usecases.validateorigin.IValidateOriginUseCase
import com.robinfood.app.usecases.validateuuid.IValidateUuidUseCase
import com.robinfood.core.dtos.transactionrequest.TransactionRequestDTO
import com.robinfood.core.dtos.transactionresponse.TransactionCreatedResponseDTO
import com.robinfood.core.enums.Result
import com.robinfood.core.mappers.toTransactionRequestEntity
import com.robinfood.repository.transactions.ITransactionRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * Implementation of ICreateTransactionUseCase
 */
class CreateTransactionUseCase(
        private val transactionRepository: ITransactionRepository,
        private val validateOriginUseCase: IValidateOriginUseCase,
        private val validateUuidUseCase: IValidateUuidUseCase
) : ICreateTransactionUseCase {

    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    override suspend fun invoke(
            token: String,
            transactionRequestDTO: TransactionRequestDTO
    ): Result<TransactionCreatedResponseDTO> {

        log.info("Execute create transaction use case {}", transactionRequestDTO)

        val transactionRequest = validateOriginUseCase.invoke(transactionRequestDTO)

        val transactionRequestVerify = validateUuidUseCase.invoke(transactionRequest)

        return transactionRepository.createTransaction(
                token,
                transactionRequestVerify.toTransactionRequestEntity()
        )
    }
}
